File: C++/34_Find_First_And_Last_Position_Of_Element_In_Sorted_Array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        //set up 2 references for binary search
        int left = 0, right = nums.size() - 1;
        
        while(left <= right) {
            //calculate mid index
            int mid = left + (right - left) / 2;
            //check if we found our target
            if(nums[mid] == target) {
                //if we did we have to adjust our boundaries for our left
                //and right pointers
                int shiftLeft = mid, shiftRight = mid;
                //keep adjusting until we reach boundary
                while(shiftLeft >= 0 && nums[shiftLeft] == target) {
                    shiftLeft--;
                }
                shiftLeft++;
                //do same thing but for right portion
                while(shiftRight < nums.size() && nums[shiftRight] == target) {
                    shiftRight++;
                }
                shiftRight--;
                //return our result
                return {shiftLeft,shiftRight};
                
            } else if(nums[mid] < target) {
                //search right half
                left = mid + 1;
            } else {
                //search left half
                right = mid - 1;
            }
        }
        return {-1,-1};
    }
};

File: Java/34_Find_First_And_Last_Position_Of_Element_In_Sorted_Array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        //set up binary search
        int left = 0, right = nums.length - 1;
        
        
        while(left <= right) {
            //get mid index
            int mid = left + (right - left) / 2;
            //check if equal
            if(nums[mid] == target) {
                //check where boundaries end for left side
                int leftSol = mid, rightSol = mid;
                while(leftSol >= 0 && nums[leftSol] == target) {
                    leftSol--;
                }
                //adjust back
                leftSol++;
                //check where boundaries end
                while(rightSol < nums.length && nums[rightSol] == target) {
                    rightSol++;
                }
                //adjust back
                rightSol--;
                //return solution
                return new int[]{leftSol,rightSol};
            } else if(nums[mid] < target) {
                //cut in half
                left = mid + 1;
            } else {
                //cut in half
                right = mid - 1;
            }
        }
        //if we exited loop we found nothing
        return new int[]{-1,-1};
    }
}
